doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,222 @@
+
+References
+----------
+
+- [1] FIPS 140-3 Standards: <https://csrc.nist.gov/projects/cryptographic-module-validation-program/fips-140-3-standards>
+- [2] Approved Security Functions: <https://csrc.nist.gov/projects/cryptographic-module-validation-program/sp-800-140-series-supplemental-information/sp800-140c>
+- [3] Approved SSP generation and Establishment methods: <https://csrc.nist.gov/projects/cryptographic-module-validation-program/sp-800-140-series-supplemental-information/sp800-140d>
+- [4] Key transitions: <https://csrc.nist.gov/pubs/sp/800/131/a/r2/final>
+- [5] FIPS 140-3 Implementation Guidance: <https://csrc.nist.gov/csrc/media/Projects/cryptographic-module-validation-program/documents/fips 140-3/FIPS 140-3 IG.pdf>
+
+Requirements
+------------
+
+The following information was extracted from the FIPS 140-3 IG [5] “2.4.C Approved Security Service Indicator”
+
+- A module must have an approved mode of operation that requires at least one service to use an approved security function (defined by [2] and [3]).
+- A FIPS 140-3 compliant module requires a built-in service indicator capable of indicating the use of approved security services
+- If a module only supports approved services in an approved manner an implicit indicator can be used (e.g. successful completion of a service is an indicator).
+- An approved algorithm is not considered to be an approved implementation if it does not have a CAVP certificate or does not include its required self-tests. (i.e. My interpretation of this is that if the CAVP certificate lists an algorithm with only a subset of key sizes, digests, and/or ciphers compared to the implementation, the differences ARE NOT APPROVED. In many places we have no restrictions on the digest or cipher selected).
+- Documentation is required to demonstrate how to use indicators for each approved cryptographic algorithm.
+- Testing is required to execute all services and verify that the indicator provides an unambiguous indication of whether the service utilizes an approved cryptographic algorithm, security function or process in an approved manner or not.
+- The Security Policy may require updates related to indicators. AWS/google have added a table in their security policy called ‘Non-Approved Algorithms not allowed in the approved mode of operation’. An example is RSA with a keysize of < 2048 bits (which has been enforced by [4]).
+
+Since any new FIPS restrictions added could possibly break existing applications
+the following additional OpenSSL requirements are also needed:
+
+- The FIPS restrictions should be able to be disabled using Configuration file options (This results in unapproved mode and requires an indicator).
+- A mechanism for logging the details of any unapproved mode operations that have been triggered (e.g. DSA Signing)
+- The FIPS restrictions should be able to be enabled/disabled per algorithm context.
+- If the per algorithm context value is not set, then the  Configuration file option is used.
```

Comment:
I have a feeling we want this the other way around.  I.e. config trumps all.

Reply:
If the configuration were to trump everything would there be any point in being able to set the value programatically?  It seems odd to allow a programatic setting iff the configuration hasn't already specified a value.

Reply:
Every time we've done one of the two, the other has been requested too.

Not allowing the configuration to override a programmatic setting is a failure for consumers who can't modify the code but can change the configuration.  That's a fairly common situation in regulated industries, the cost and time to re-certify can be massive.

Reply:
Not sure I understand what you are saying.. If they cant set per ctx via code then there is no priority issue. If you are talking about edge cases where we have to disable an option internally in code (for a protocol for example) then that is an override that doesnt care what the config wants.

Reply:
IMO it should be like this for most cases. Possibly there are some exceptions where it would make sense to do it otherwise so I am saying most not all.

1. By default the check is enforced.
2. If config file says not enforced, it is not enforced.
3. If API call says not enforced, it is not enforced.

I.e. this is an OR operation between config and API call for non-enforcement.

Reply:
For some extraordinarily _inconvenient_ requirements where they would break _everyone_ we might want to have also this:

1. By default this check is unenforced.
2. If config file says enforced, it is enforced.
3. If API call says not enforced, it is not enforced.

I.e., for checks unenforced by default we allow overriding the default first by config file and then by the API call.

Reply:
I am not sure I understand why this is needed, you already have the option on turning off the fips property.
Is this because you want the fips provider to still be doing the unapproved operatiosn but just return an indicator that the operation was unapproved?

If that is the case, why not simply allow to set a global fips=2 property or some such, that indicates what you want?